src/core/coap/coap_message.cpp

```diff
@@ -373,18 +373,22 @@ Error Message::SetPayloadMarker(void)
 
 Error Message::ParseHeader(void)
 {
-    Error            error = kErrorNone;
+    Error            error  = kErrorNone;
+    uint16_t         offset = GetOffset();
     Option::Iterator iterator;
 
     OT_ASSERT(GetReserved() >=
               sizeof(HelpData) + static_cast<size_t>((reinterpret_cast<uint8_t *>(&GetHelpData()) - GetFirstData())));
 
     GetHelpData().Clear();
 
-    GetHelpData().mHeaderOffset = GetOffset();
-    IgnoreError(Read(GetHelpData().mHeaderOffset, GetHelpData().mHeader));
+    GetHelpData().mHeaderOffset = offset;
+
+    SuccessOrExit(error = Read(offset, &GetHelpData().mHeader, kMinHeaderLength));
+    offset += kMinHeaderLength;
 
     VerifyOrExit(GetTokenLength() <= kMaxTokenLength, error = kErrorParse);
+    SuccessOrExit(error = Read(offset, GetHelpData().mHeader.mToken, GetTokenLength()));
```

Comment:
In the new push, changed this from `&GetHelpData().mHeader.mToken` (which is technically correct but can be interpreted wrongly) to simply `GetHelpData().mHeader.mToken` since `mToken` is an array